nohup/nohup.1
.\"
.\" Sccsid @(#)nohup.1	1.6 (gritter) 1/24/05
.\" Parts taken from nice(1), Unix 7th edition:
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\"
.\" SPDX-Licence-Identifier: Caldera
.\"
.TH NOHUP 1 "1/24/05" "Heirloom Toolchest" "User Commands"
.SH NAME
nohup \- run a command immune to hangups
.SH SYNOPSIS
\fBnohup\fI utility \fR[\fIarguments\fR\ ...]
.SH DESCRIPTION
The
.B nohup
command executes
.I utility
immune to hangup signals,
and, for
.B /usr/5bin/nohup
and
.BR /usr/5bin/s42/nohup ,
also immune to quit signals.
If standard output is not re-directed by the user to a non-tty file,
it will be appended to the file `\fInohup.out\fR'.
If \fInohup.out\fR is not writable
in the current directory,
output is appended to
.IR $HOME/nohup.out .
If these files do not exist,
they are created with mode 666 for
.B /usr/5bin/nohup
and
.B /usr/5bin/s42/nohup
or with mode 600 for
.B /usr/5bin/posix/nohup
and
.BR /usr/5bin/posix2001/nohup .
If standard error is not re-directed by the user to a non-tty file,
it will be sent to the same file as standard output.
.SH "ENVIRONMENT VARIABLES"
.TP
.B PATH
Used to locate
.IR utility .
.SH "SEE ALSO"
chmod(1),
nice(1),
sh(1),
signal(2),
execvp(3),
isatty(3)


CHANGES
Release ...
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants. 

Release 230503 
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin. 

Release 230120
* Notes for the library directories if linking against the GNU C library.
* rm: Consistency within standards: the SUS program shouldn't have the POSIX 1003.1(2008) fix, just the main program.
* Made $(ROOT) not hardcoded anymore.
* Cleaned up more licence headers (if my estimations are correct, we've got rid of 3.255 lines!).
* Small update on README.md --- although it's now less readable on plain-text readers such as more or pg.

Release 230107
* rm: Added support for the POSIX 1003.1(2008).
* chroot: Added manual page.
* awk: Added stdint.h to awk.g.y.
* Now it has a bigger range of compatibility with GNU C library versions.

Release 221025
* Made notes clearer about building Heirloom on modern versions of the GNU C library.
* tar: Support for zstd and xz.

Release 220821
* [NEW] chroot: New implementation of chroot, based on Solaris 2.5 payloads.
* true, false: Rewritten in C.
* yes: Rewritten and adapted to ANSI C.
* "libcommon/vpfmt.c": Adapted to ANSI C.
* Now many of the utilities can be built in the musl C library and GNU C library (dynamically).

Release 070715
* cpio: A "-P" option has been introduced that allows to specify attributes
  like the file owner when creating an archive. With it, unprivileged users
  can set arbitrary attributes for files within an archive.
* diff: Made it compile on CentOS 5.
* Static linking against glibc is now possible since "optopt" is no longer
  a common symbol, and there is no conflict anymore with "getopt.o".
* cpio, tar: A compilation problem on some Linux systems has been fixed.
* ul: A compilation problem for NetBSD and other systems that use termcap
  instead of curses has been fixed (patch by Aleksey Cheusov).
* bfs: An OpenSolaris-derived manual page is now included.

Release 070227
* The directory for development utilities, by default "/usr/ccs/bin", is
  now configurable in "build/mk.config" . It is expected that the utilities
  from the "Heirloom Development Tools" package are installed in it.
* getconf: "/usr/ccs/bin" is now included in the default path as printed
  with the "PATH" argument.
* what: Has been replaced by an OpenSolaris-derived implementation.
* tsort: Is now installed in "/usr/ccs/bin".
* OpenSolaris-derived manual pages are now included for bdiff, newform,
  and fspec(5).
* psrinfo: Fixed the computation of the "on-line since" time for Linux 2.6.

nohup/nohup.c
/*
 * nohup - run a command immune to hangups
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, August 2002.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * SPDX-Licence-Identifier: Zlib
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
#ifdef	SUS
static const char sccsid[] USED = "@(#)nohup_sus.sl	1.11 (gritter) 5/29/05";
#else
static const char sccsid[] USED = "@(#)nohup.sl	1.11 (gritter) 5/29/05";
#endif

#include	<sys/types.h>
#include	<sys/stat.h>
#include	<fcntl.h>
#include	<unistd.h>
#include	<stdio.h>
#include	<string.h>
#include	<stdlib.h>
#include	<errno.h>
#include	<signal.h>
#include	<libgen.h>
#include	<limits.h>

static char	*progname;		/* argv[0] to main() */
static int	errorfd = 2;

static void
fatal(int eno, const char *string, int val)
{
	char	buf[PATH_MAX];

	snprintf(buf, sizeof buf, "%s: %s: %s\n", progname, string,
				strerror(eno));
	write(errorfd, buf, strlen(buf));
	exit(val);
}

/*
 * Memory allocation with check.
 */
static void *
srealloc(void *vp, size_t nbytes)
{
	void *p;

	if ((p = realloc(vp, nbytes)) == NULL) {
		write(2, "Out of memory\n", 14);
		exit(077);
	}
	return p;
}

static void *
smalloc(size_t nbytes)
{
	return srealloc(NULL, nbytes);
}

static void
usage(void)
{
	fprintf(stderr, "usage: %s command arg ...\n", progname);
	exit(0177);
}

static void
success(const char *fn)
{
	fprintf(stderr, "Sending output to %s\n", fn);
}

static void
failed(const char *fn)
{
	fprintf(stderr, "%s: cannot open/create %s\n", progname, fn);
	exit(0177);
}

static void
create_nohup_out(void)
{
	char	*nohup_out = "nohup.out";
	char	*fn;
	int	fd;
#ifdef	SUS
	const mode_t	mode = S_IRUSR | S_IWUSR;
#else
	const mode_t	mode = 0666;
#endif

	fn = nohup_out;
	if ((fd = open(fn, O_WRONLY|O_CREAT|O_APPEND, mode)) < 0) {
		char	*home;

		if ((home = getenv("HOME")) != NULL) {
			char	*cp, *cq;

			cp = fn = smalloc(strlen(home) + strlen(nohup_out) + 2);
			for (cq = home; *cq; cq++)
				*cp++ = *cq;
			*cp++ = '/';
			cq = nohup_out;
			while ((*cp++ = *cq++) != '\0');
			if ((fd = open(fn, O_WRONLY|O_CREAT|O_APPEND,
							mode)) < 0)
				failed(fn);
		} else
			failed(fn);
	}
	success(fn);
	if (dup2(fd, 1) < 0)
		fatal(errno, "dup2", 0177);
	close(fd);
}

int
main(int argc, char **argv)
{
	int i;

	progname = basename(argv[0]);
	if (argc > 1 && argv[1][0] == '-' && argv[1][1] == '-'
			&& argv[1][2] == '\0')
		argv++, argc--;
	if (argc <= 1)
		usage();
	if (isatty(1))
		create_nohup_out();
	errorfd = dup(2);
	fcntl(errorfd, F_SETFD, FD_CLOEXEC);
	if (isatty(2))
		if (dup2(1, 2) < 0)
			fatal(errno, "dup2", 0177);
	signal(SIGHUP, SIG_IGN);
#ifndef	SUS
	signal(SIGQUIT, SIG_IGN);
#endif
	execvp(argv[1], &argv[1]);
	i = errno;
	fatal(i, argv[1], i == ENOENT ? 0177 : 0176);
	/*NOTREACHED*/
	return 0177;
}
